source/common/common/token_bucket_impl.h

```diff
@@ -35,4 +35,36 @@ class TokenBucketImpl : public TokenBucket {
   TimeSource& time_source_;
 };
 
+/**
+ * Atomic token bucket. This class is thread-safe.
+ */
+class AtomicTokenBucketImpl {
+public:
+  /**
+   * @param max_tokens supplies the maximum number of tokens in the bucket.
+   * @param time_source supplies the time source.
+   * @param fill_rate supplies the number of tokens that will return to the bucket on each second.
+   * The default is 1.
+   * @param init_fill supplies whether the bucket should be initialized with max_tokens.
+   */
+  explicit AtomicTokenBucketImpl(uint64_t max_tokens, TimeSource& time_source,
+                                 double fill_rate = 1.0, bool init_fill = true);
+
+  /**
+   * Consume tokens from the bucket.
+   * @param tokens supplies the number of tokens to consume.
+   * @param allow_partial supplies whether partial consumption is allowed.
+   * @param factor supplies the factor to multiply the fill rate by. This is used to change the fill
+   * rate based on the requirement of the caller.
+   * @return the number of tokens consumed.
+   */
+  uint64_t consume(uint64_t tokens, bool allow_partial, double factor = 1.0);
+
+private:
+  const double max_tokens_;
+  const double fill_rate_;
+  std::atomic<MonotonicTime> time_in_nanoseconds_{};
```

Comment:
Any idea what the physical size of MonotonicTime is? Since it's not representing a true time with large fill rate, is there any concern about overflow?

Reply:
8 bytes. It's a timestamp of actual time.
The only change to overflow is that: the max_token is set to `(2024-1970)` and the fill rate is set to the min fill rate (1 token per year), and the initial fill is set to true. I will add a check to this corner case.

Reply:
Hah, even overflow is OK anyway. minus number also could be used here anyway.